src/pages/api/login.js

```diff
@@ -0,0 +1,49 @@
+import { PrismaClient } from '@prisma/client';
+import { setCookie } from 'cookies-next';
+
+const prisma = new PrismaClient();
+
+export default async function login(req, res) {
+    const sha_js = require('sha.js');
+    let token;
+    if (req.method === 'POST') {
+        const username = req.body.username;
+        const password = req.body.password;
+        let pass_hash = sha_js('sha256').update(password).digest('hex');
+        const user = await prisma.user.findMany({
+            where: {
+                username: username,
+                password_hash: pass_hash,
+            },
+        });
+        if (user.length === 0) {
+            res.failure = true;
+            res.status(200).redirect('/login?status=error');
+        } else {
+            const user_token = await prisma.user_token.findMany({
+                where: {
+                    userId: user[0].id,
+                },
+            });
+            if (
+                user_token.length === 0 ||
+                user_token[0].expires.valueOf() < Date.now().valueOf()
+            ) {
+                token = require('random-token')(32);
```

Comment:
again, you don't need to generate random tokens within js, prisma will do it automatically, if you add code to your schema, that i suggested

Reply:
this way is better, because i don't need to fetch db again